asto/asto-s3/src/main/java/com/artipie/asto/s3/MultipartUpload.java

```diff
@@ -78,34 +77,30 @@ final class MultipartUpload {
 
     /**
      * Uploads all content by parts.
+     * Note that content part must be at least MultipartUpload.MIN_PART_SIZE except last part.
      *
      * @param content Object content to be uploaded in parts.
      * @return Completion stage which is completed when responses received from S3 for all parts.
      */
     public CompletionStage<Void> upload(final Content content) {
         final AtomicInteger counter = new AtomicInteger();
-        return Flowable.fromPublisher(content)
-            .concatMap(
+        return new Merging(MultipartUpload.MIN_PART_SIZE).mergeFlow(
+            Flowable.fromPublisher(content).concatMap(
                 buffer -> Flowable.fromPublisher(
                     new Splitting(buffer, MultipartUpload.MIN_PART_SIZE).publisher()
                 )
-            ).map(
-                chunk -> {
-                    final int pnum = counter.incrementAndGet();
-                    return this.uploadPart(
-                        pnum,
-                        Flowable.just(chunk)
-                    ).thenAccept(
-                        response -> this.parts.add(
-                            new UploadedPart(pnum, response.eTag())
-                        )
-                    );
-                }
-            ).reduce(
-                CompletableFuture.allOf(),
-                (acc, stage) -> acc.thenCompose(o -> stage)
-            ).to(SingleInterop.get())
-            .thenCompose(Function.identity());
+            )
+        ).doOnNext(payload -> {
+            final int pnum = counter.incrementAndGet();
+            this.uploadPart(
+                pnum,
+                Flowable.just(payload)
+            ).thenAccept(
+                response -> this.parts.add(
+                    new UploadedPart(pnum, response.eTag())
+                )
+            ).toCompletableFuture().join();
```

Comment:
Why do we block here?

Reply:
@dgarus , here we send one request with chunk data at time. We shouldn't send all chunks/requests in parallel, since for big files it will overflow request pool of the AWS S3 client, and also it may overload limits for S3 server or even firewall. I'll add this note to the method comment, since its important.